Express/src/services/collectionService.ts

```diff
@@ -0,0 +1,168 @@
+import { BookWithAuthorsAndGenres } from "../../types/BookAuthorGenre";
+import { BookWithAuthorsAndGenres } from "../../types/BookAuthorGenre";
+import * as dataAcess from "../data-access/collectionDataAccess";
+
+export async function fetchPublicCollections() {
+  const publicCollections = dataAcess.getPublicCollections();
+  if (!publicCollections.length) {
+    throw { status: 404, message: "No public collections found" };
+  }
+  return publicCollections;
+}
+
+export async function fetchUserCollections(userId: string) {
+  const userCollections = await dataAcess.getUserCollections(userId);
+  if (!userCollections.length) {
+    throw { status: 404, message: "No collection found" };
+  }
+  return userCollections;
+}
+
+export async function fetchCollectionDetails(
+  collectionId: number,
+  userId?: string
+) {
+  const collection = await dataAccess.getCollectionByID(collectionId);
+  if (collection == null) {
+    throw { status: 404, message: "Collection not found" };
+  }
+
+  const isOwner = collection.userId == userId;
+  if (!collection.public && isOwner === false) {
+    throw { status: 403, message: "Forbidden: Access denied" };
+  }
+
+  const booksInCollection = await dataAccess.getBooksInCollection(collectionId);
+
+  const booksWithDetails: BookWithAuthorsAndGenres[] = [];
+  for (const book of booksInCollection) {
+    const [authors, genres] = Promise.all([
+      dataAccess.getBookAuthors(book.id),
+      dataAccess.getBookGenres(book.id),
+    ]);
+
+    booksWithDetails.push({
+      ...book,
+      authors: authors.map((author) => author.names),
+      genres: genres.map((genre) => genre.genre),
+    });
+  }
```

Comment:
Šia Promise.all logiką galima iškelti iš ciklo ir panaudoti map funkciją. Šis atnaujinimas pagerina našumą ir skaitomumą, nes naudojant Promise.all vienu metu paimami autoriai ir žanrai kiekvienai knygai, esančiai booksInCollection.
```suggestion
  const booksWithDetails: BookWithAuthorsAndGenres[] = await Promise.all(
    booksInCollection.map(async (book) => {
      const [authors, genres] = await Promise.all([
        dataAccess.getBookAuthors(book.id),
        dataAccess.getBookGenres(book.id),
      ]);
      return {
        ...book,
        authors: authors.map((author) => author.name),
        genres: genres.map((genre) => genre.genre),
      };
    })
  );
```
Tokiu būdu visų knygų duomenis gauname greičiau ir švaresniu, lengviau skaitomu kodu.